allen-cell-animated/nucmorph-colorizer
# fix: Fix crash from `ax.dtick` NaN error

Problem
=======
*Estimated size: small, 10-15 minutes*

Closes #291, a crash with the error message `ax.dtick error: NaN` on certain datasets.

I think what was happening was that Plotly can handle _some_ `Infinity` or `NaN` values in input data, but fails when calculating the range of axis values to show if _all_ the values are `Infinity` or `NaN`. 

This happens in some datasets on certain frame + feature combinations, such as in the updated nucmorph datasets. The **time\*** feature fails on the first frame, when values are undefined for all the cells currently in frame.

(\*representing the age of a cell since mitosis)

I also opened up a related issue here: [#329, NaN in feature data is shown as Infinity once loaded.](https://github.com/allen-cell-animated/nucmorph-colorizer/issues/329)

Solution
========
- Unrepresentable `NaN` and `Infinity` values are now removed from scatter plot data before rendering.
- Adds a warning banner for when the plot rendering fails for any reason, and clears the plot.

## Type of change
* Bug fix (non-breaking change which fixes an issue)

Steps to Verify:
----------------
(Note: these datasets are pretty big, so they may take up to 10 seconds to load.)
1. Open the bugged build: https://dev-aics-dtp-001.int.allencell.org/nucmorph-colorizer/dist/index.html#/viewer?collection=https%3A%2F%2Fdev-aics-dtp-001.int.allencell.org%2Fdan-data%2Fcolorizer%2Fdata%2Fnucmorph-apr2024%2Fcollection.json&dataset=Mama+Bear&feature=time&tab=scatter_plot&scatter-range=all&scatter-x=time&scatter-y=volume
2. Note that all cells on the current frame have a value of `Infinity`.
3. Go to the scatter plot tab and change the range from **All time** to **Current frame**. The UI will crash.
4. Now open the preview link and repeat the above instructions. No error should occur. https://allen-cell-animated.github.io/nucmorph-colorizer/pr-preview/pr-330/#/viewer?collection=https%3A%2F%2Fdev-aics-dtp-001.int.allencell.org%2Fdan-data%2Fcolorizer%2Fdata%2Fnucmorph-apr2024%2Fcollection.json&dataset=Mama+Bear&feature=time&tab=scatter_plot&scatter-range=all&scatter-x=time&scatter-y=volume


## Screenshots:
**Video showing off the error banner (🔊):**

https://github.com/allen-cell-animated/nucmorph-colorizer/assets/30200665/b8663610-bf0e-4472-bb94-d4fcf94ce9a5

![image](https://github.com/allen-cell-animated/nucmorph-colorizer/assets/30200665/a4cc18ff-96b3-43f3-9415-07f4ceac587e)
